### ISSUE TITLE
Error 0x9 operation failed

### ISSUE THREAD

**Issue by bfg01** [stefanberger/swtpm#747]
On this system I have QEMU 6 and swtpm 0.5.3 (yes, old because this is the only version distro ships).

So I start swtpm with:
swtpm socket -t --tpm2 --tpmstate dir=tpm/tpm0 --ctrl type=unixio,path=tpm/tpm0/swtpm-sock --log level=20 -d

And use these options for QEMU:
-chardev socket,id=chrtpm,path=tpm/tpm0/swtpm-sock -tpmdev emulator,id=tpm0,chardev=chrtpm -device tpm-tis,tpmdev=tpm0

Try to run QEMU and I get this error:
qemu-system-x86_64: tpm-emulator: TPM result for CMD_INIT: 0x9 operation failed
Even the swtpm process is deleted

Directory /var/lib/swtpm-localca has ownership/permissions tss:root drwxr-x---
There are no logs at all inside /var/log/swtpm/libvirt/qemu/

Could you help a bit please?
Thanks.

**Comment by stefanberger** [comment]
Can you/did you follow the instructions here (as root)? https://github.com/qemu/qemu/blob/master/docs/specs/tpm.rst#the-qemu-tpm-emulator-device

**Comment by bfg01** [comment]
Of course I did, as they're reflected in what I posted.

Just with two details:
Out of the only 2 options SeaBIOS or OVMF, I'm using OVMF
I'm not running neither QEMU nor swtpm as root; I really prefer to avoid this as much as possible, and even the QEMU docs advise against it

I'm on openSUSE Leap 15.4, by the way.

**Comment by stefanberger** [comment]
> Of course I did, as they're reflected in what I posted.

You did not state whether you ran it as root. I would try running it as root first and see whether this works. Then you can try to run it as non-root afterwards.

A detailed description of all your steps would be helpful, so one can recreate the issue step  by step elsewhere.

Can you clean the state directory where swtpm writes its state into?

> Directory /var/lib/swtpm-localca has ownership/permissions tss:root drwxr-x---

Only `swtpm_setup` would use this directory, `swtpm` doesn't use. It shoudn't matter in what you are trying to do.

**Comment by bfg01** [comment]
> A detailed description of all your steps would be helpful, so one can recreate the issue step by step elsewhere.
Ok, agrede. Aside trying with root, what other particular information would you need?

> Can you clean the state directory where swtpm writes its state into?
Do you mean the "tmp/tmp0" directory? (Which full path is ~/tmp/tmp0). I did delete everything inside it.

**Comment by bfg01** [comment]
Running as root both QEMU and swtpm didn't work, still the exact same issue.

Just for some more detail, I use OVMF like this:
-drive if=pflash,format=raw,readonly=on,file=uefi/ovmf-x86_64-code.bin \
-drive if=pflash,format=raw,file=uefi/ovmf-x86_64-vars.bin \
All the rest of options are just for describing VM's hardware, which should be unrelated.

I tried something else: enabling debug with QEMU option -d trace:tpm_\* . Still no dice:
```tpm_emulator_set_buffer_size buffer size: 4096, min: 2808, max: 4096
tpm_emulator_startup_tpm_resume is_resume: 0, buffer size: 4096
tpm_emulator_set_buffer_size buffer size: 4096, min: 2808, max: 4096
qemu-system-x86_64: tpm-emulator: TPM result for CMD_INIT: 0x9 operation failed```
I also tried with 4 Mb ("4m") versions of OVMF.

Only option left was trying building version 0.7.3 from source.
It was not easy: with all due respect, but the wiki page lacks correct info regarding the required dependencies to build, at very least for the openSUSE case. So, for Leap 15.4 this was it in the end:
```zypper in socat net-tools-deprecated libtasn1-devel gnutls-devel libseccomp-devel json-glib-devel gcc make automake autoconf libtool libtpms-devel libopenssl-*-devel expect```
Using wildcard with openssl because it depends on which version is currently available.

But luckily, it ended up working at last. No more error and QEMU seemingly finally ran.

Now final doubts:
---What's the -t option actually for? I saw that, with or without it, once QEMU exits the socket in ~/tmp/tmp0 is deleted anyways (which doesn't happen with version 0.5.3!), as well as the swtpm process terminated.
---What are the .lock and tpm2-00.permall files for, which remain there after exiting QEMU?

Thanks very much again. Not closing this issue yet since I'd still like to diagnose what's wrong with version 0.5.3 if possible, unless author considers it better.

**Comment by stefanberger** [comment]
> Running as root both QEMU and swtpm didn't work, still the exact same issue.

I cannot recreate this issue with QEMU 6.2 and swtpm 0.5.3. Starting a QEMU VM as described [here](https://github.com/qemu/qemu/blob/master/docs/specs/tpm.rst#the-qemu-tpm-emulator-device) works fine for me.

> 
> Just for some more detail, I use OVMF like this: -drive if=pflash,format=raw,readonly=on,file=uefi/ovmf-x86_64-code.bin -drive if=pflash,format=raw,file=uefi/ovmf-x86_64-vars.bin All the rest of options are just for describing VM's hardware, which should be unrelated.
> 
> I tried something else: enabling debug with QEMU option -d trace:tpm_* . Still no dice:
> 
> ```
> tpm_emulator_set_buffer_size buffer size: 4096, min: 2808, max: 4096
> tpm_emulator_startup_tpm_resume is_resume: 0, buffer size: 4096
> tpm_emulator_set_buffer_size buffer size: 4096, min: 2808, max: 4096
> qemu-system-x86_64: tpm-emulator: TPM result for CMD_INIT: 0x9 operation failed
> ```

I don't know where this may be coming from and why it would work with a later version of swtpm.

> 
> I also tried with 4 Mb ("4m") versions of OVMF.

That shouldn't influence it at all. Neither usage of SeaBIOS or EDK2 should have any influence on what you seem to be seeing. Also, whether it's QEMU 6.2 or QEMU 6.0 shouldn't make a difference.

> 
> Only option left was trying building version 0.7.3 from source. It was not easy: with all due respect, but the wiki page lacks correct info regarding the required dependencies to build, at very least for the openSUSE case. So, for Leap 15.4 this was it in the end:
> 
> ```
> zypper in socat net-tools-deprecated libtasn1-devel gnutls-devel libseccomp-devel json-glib-devel gcc make automake autoconf libtool libtpms-devel libopenssl-*-devel expect
> ```

It looks like the command line may have been cut off. I will take what I have been using in this file [here](https://github.com/stefanberger/swtpm-distro-compile/blob/master/test.opensuse-tumbleweed#L6-L8) and put it on the wiki. Thanks though. I have used this on Travis for testing now for quite some time with SuSE Tumbleweed, so it should be the correct packages and openssl-devel seems to be the package to use for getting the openssl headers.

> 
> Using wildcard with openssl because it depends on which version is currently available.
> 
> But luckily, it ended up working at last. No more error and QEMU seemingly finally ran.
> 
> Now final doubts: ---What's the -t option actually for? I saw that, with or without it, once QEMU exits the socket in ~/tmp/tmp0 is deleted anyways (which doesn't happen with version 0.5.3!), as well as the swtpm process terminated. ---What are the .lock and tpm2-00.permall files for, which remain there after exiting QEMU?


From the man page:
```
-t|--terminate

    Terminate the TPM after the client has closed the connection.
```

This terminates swtpm if QEMU didn't have the chance to send it a CMD_SHUTDOWN upon termination. So abrupt termination of QEMU (crash, killed by signal) terminates swtpm as well.

TPMs are stateful and tpm2-00.permall is the file where the TPM stores its state in. The .lock file is for swtpm (directory storage backend) to get an exclusive lock on the directory where the state file is in preventing another instance of swtpm from accessing the statefile at the same time.

> 
> Thanks very much again. Not closing this issue yet since I'd still like to diagnose what's wrong with version 0.5.3 if possible, unless author considers it better otherwise.

I don't see the error you are seeing with 0.5.3 and don't know where it may come from...

**Comment by bfg01** [comment]
Yes, you're right. I think I have been too evasive, and I apologize for that and for having you waste time...

But now, I think I was finally able to replicate the problem in a more faithful way:

Setting: openSUSE Leap 15.4 with KDE, QEMU 6.2, swtpm 0.5.3
Swtpm installed via "zypper in swtpm"; I think these are the related links just in case:
https://build.opensuse.org/repositories/openSUSE:Leap:15.4/swtpm
https://build.opensuse.org/package/show/SUSE:SLE-15-SP3:Update/swtpm
https://download.opensuse.org/distribution/leap/15.4/repo/oss/x86_64/swtpm-0.5.3-150300.3.3.1.x86_64.rpm.mirrorlist

Start swtpm with
`swtpm socket -t --tpm2 --tpmstate dir=~/tpm --ctrl type=unixio,path=~/tpm/swtpm-sock --log level=20 -d`

Start QEMU with
```
qemu-system-x86_64 -display none -accel kvm \
  -m 1024 -boot d -bios bios-256k.bin -boot menu=on \
  -chardev socket,id=chrtpm,path=~/tpm/swtpm-sock \
  -tpmdev emulator,id=tpm0,chardev=chrtpm \
  -device tpm-tis,tpmdev=tpm0 \
  -drive file=<name>.iso,index=0,media=cdrom
```

Where <name> is a number of Linux live systems I tried, such as Slitaz, Fedora, AntiX, openSUSE itself... Also tried with a Win10 ISO. I don't have a "test.img" to test with...

With all this I do get the error.
Also, other symptoms are: the swtpm process is terminated with or without the "-t" option, and the socket inside ~/tpm is never deleted.

With swtpm 0.7.3 built from source the problems are gone, as already stated.

Truth to be told I haven't tried with Tumbleweed's 0.7.3... Also, could it be possible that you tried with 0.5.3 built from source instead of from SUSE's repositories?

I hope this time it's better. Thanks and sorry again.

**Comment by stefanberger** [comment]
I cannot use `~/tmp` as the path at all. I have to use `/home/stefanb/tmp`. I works for me just as fine as non-root user. I am not using SuSE but Fedora as the host but with swtpm v0.5.3 and QEMU 6.2.

In one terminal as non-root user:   [ without --tpm2 this will start a TPM 1.2 ]

```
$ rm -rf ~/tmp/
$ mkdir ~/tmp
$ swtpm socket --tpmstate dir=~/tmp   --ctrl type=unixio,path=~/tmp/swtpm-sock   --log level=20
swtpm: Could not open UnixIO socket: No such file or directory
$ swtpm socket --tpmstate dir=/home/stefanb/tmp   --ctrl type=unixio,path=/home/stefanb/tmp/swtpm-sock   --log level=20
```

In other terminal as non-root user:

```
qemu-system-x86_64 -display sdl -accel kvm   -m 1024 -boot d -bios bios-256k.bin -boot menu=on   -chardev socket,id=chrtpm,path=/home/stefanb/tmp/swtpm-sock   -tpmdev emulator,id=tpm0,chardev=chrtpm   -device tpm-tis,tpmdev=tpm0
```


I can now enter the SeaBIOS TPM menu and enable and disable the TPM and it seems to work just fine.

**Comment by bfg01** [comment]
Yes, sorry, I forgot I should use either full explicit path or at least "$HOME".

I tried again, this time without --tpm2 and without the final -drive option I used in QEMU, just like you suggest. Again, both as normal user and as root.

Exact same result.

And, if you always tested in Fedora since the beginning and not in SUSE, and given the compiled 0.7.3 is actually working here, then I think all of this would narrow to just one possibility: SUSE's 0.5.3 package is simply broken.

If I'm right this should go to SUSE's bugzilla. So, unless you needed more tests from my side, could I leave the closing of this issue to your criteria?

Thanks very much for all your help.

**Comment by stefanberger** [comment]
Can you try to build v0.5.3 from the sources and try again and if that also doesn't work the tip of the stable-0.5 branch? I just looked at their source rpm and didn't find anything that looks particularly bad for their v0.5.3.

**Comment by stefanberger** [comment]
> 
> Ok, fun is over: is it that it's an issue _exclusive_ only to my trash rig? Right now I don't have any other PC where to test...

I don't know the answer but it's odd that I cannot recreate the issue.

**Comment by stefanberger** [comment]
I just repeated the test with OpenSuSE 15.3 and could not recreate your issue.

**Comment by stefanberger** [comment]
At this point I think we should either close this issue, Otherwise you should provide the commands you run on your systems so that this issue can be recreated by just copying and pasting these commands into a terminal.

**Comment by bfg01** [comment]
Until then I had tried in an old laptop Asus X550. Now I tried with another different rig, Acer A315, from this year. I still get the exact error, by following exactly this:
```
swtpm socket -t -d --tpm2 --tpmstate dir="$HOME"/tpm --ctrl type=unixio,path="$HOME"/tpm/swtpm-sock --log level=20

qemu-system-x86_64 -accel kvm -m 1024 -boot d -bios /usr/share/qemu/bios-256k.bin -boot menu=on -display gtk -chardev socket,id=chrtpm,path="$HOME"/tpm/swtpm-sock -tpmdev emulator,id=tpm0,chardev=chrtpm -device tpm-tis,tpmdev=tpm0
```

And still getting the same error:
`tpm-emulator: TPM result for CMD_INIT: 0x9 operation failed`

However, as already mentioned, I'm getting this error only with Leap's swtpm version, **not** if building *any* swtpm version from source.

All this whole time I have been testing with specific version Leap 15.4. However, I did give SLES a quick testing, and indeed no issue there (though I didn't try 15.3 anymore...).

Out of curiosity, was there a reason you chose to try 15.3 instead of latest 15.4? Could it be it's true what some say, that 15.4 was actually released with a number of bugs? By any chance, do you think you could give specifically 15.4 a try?

Whether you decide yes or no, thanks very much for your support, and sorry for being annoying. I think this is the last thing I'm reporting.
Thanks.

**Comment by stefanberger** [comment]
> Out of curiosity, was there a reason you chose to try 15.3 instead of latest 15.4? Could it be it's true what some say, that 15.4 was actually released with a number of bugs? By any chance, do you think you could give specifically 15.4 a try?

I had found Leap 15.3 VM image on one of my machines. Besides that :

libtpms seems to be the same version for LEAP 15.3 and 15.4: https://software.opensuse.org/package/libtpms

Both seems to use libpms0-0.8.2-3.3.1.x86_64. 

And you said you recompiled swtpm. The same version of swtpm is being used by 15.3 and 15.4: http://rpmfind.net/linux/rpm2html/search.php?query=swtpm&submit=Search+...&system=OpenSuSE&arch=

I have swtpm-0.5.3-150300.3.3.1.x86_64 installed on my Leap 15.3.

I have some trust in these rpm search engines...  and following my update now from 15.3 to 15.4 neither libtpms package nor swtpm packages have changed. Repeating the tests now from before on 15.4 still leads to the same results. swtpm displays lots of commands and responses (--log level=20).

**Comment by bfg01** [comment]
Yes, I know. When mentioning "rebuilt any version" of course this included 0.5.3 itself. Built from source it works; Leap 15.4's rpm doesn't.

If upgrading from 15.3 to 15.4 didn't give you the described issue, then I can only thing about two last possibilities:
1) Only reproducible with a **clean** Leap 15.4 installation from scratch, which was my case the whole time
2) Luck really just hates me by giving me two different rigs with the issue

Thanks yet again.
